mlir/include/mlir/Analysis/Presburger/Fraction.h

```diff
@@ -0,0 +1,101 @@
+//===- Simplex.h - MLIR Fraction Class --------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is a simple class to represent fractions. It supports multiplication,
+// comparison, floor, and ceiling operations.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_ANALYSIS_PRESBURGER_FRACTION_H
+#define MLIR_ANALYSIS_PRESBURGER_FRACTION_H
+
+namespace mlir {
+
+template <typename T>
+struct Fraction {
+  Fraction() : num(0), den(1) {}
+  Fraction(T oNum, T oDen) : num(oNum), den(oDen) {
+    if (den < 0) {
+      num = -num;
+      den = -den;
+    }
+  }
+  T num, den;
+};
+
+template <typename T>
+int compare(const Fraction<T> &x, const Fraction<T> &y) {
+  auto diff = x.num * y.den - y.num * x.den;
+  if (diff > 0)
+    return +1;
+  if (diff < 0)
+    return -1;
+  return 0;
+}
+
+template <typename T>
```

Comment:
Why not: return (f.num >= 0) ? (fnum/f.den) : ((f.num / f.den) - (f.num % f.den != 0)); 
it saves 3 lines.